src/hu_stuff.c

```diff
@@ -885,6 +885,449 @@ static int HU_top(int i, int idx1, int top1)
+static void HU_widget_build_keys (void)
+{
+  int i, k;
+  char *s;
+
+  hud_keysstr[4] = '\0';    //jff 3/7/98 make sure deleted keys go away
+
+  //jff add case for graphic key display
+  if (!deathmatch && hud_graph_keys)
+  {
+    i = 0;
+    hud_gkeysstr[i] = '\0'; //jff 3/7/98 init graphic keys widget string
+
+    // build text string whose characters call out graphic keys from fontk
+    for (k = 0; k < 6; k++)
+    {
+      // skip keys not possessed
+      if (!plr->cards[k])
+        continue;
+
+      hud_gkeysstr[i++] = '!'+k; // key number plus '!' is char for key
+      hud_gkeysstr[i++] = ' ';   // spacing
+      hud_gkeysstr[i++] = ' ';
+    }
+    hud_gkeysstr[i] = '\0';
+  }
+  else // not possible in current code, unless deathmatching,
+  {
+    i = 4;
+
+    hud_keysstr[i] = '\0'; //jff 3/7/98 make sure deleted keys go away
+
+    // if deathmatch, build string showing top four frag counts
+    if (deathmatch) //jff 3/17/98 show frags, not keys, in deathmatch
+    {
+      int top1 = -999, top2 = -999, top3 = -999, top4 = -999;
+      int idx1 = -1, idx2 = -1, idx3 = -1, idx4 = -1;
+      int fragcount, m;
+
+      // scan thru players
+      for (k = 0; k < MAXPLAYERS; k++)
+      {
+        // skip players not in game
+        if (!playeringame[k])
+          continue;
+
+        fragcount = 0;
+
+        // compute number of times they've fragged each player
+        // minus number of times they've been fragged by them
+        for (m = 0; m < MAXPLAYERS; m++)
+        {
+          if (!playeringame[m])
+            continue;
+          fragcount += (m!=k) ? players[k].frags[m] : -players[k].frags[m];
+        }
+
+        // very primitive sort of frags to find top four
+        if (fragcount > top1)
+        {
+          top4 = top3; top3 = top2; top2 = top1; top1 = fragcount;
+          idx4 = idx3; idx3 = idx2; idx2 = idx1; idx1 = k;
+        }
+        else if (fragcount > top2)
+        {
+          top4 = top3; top3 = top2; top2 = fragcount;
+          idx4 = idx3; idx3 = idx2; idx2 = k;
+        }
+        else if (fragcount > top3)
+        {
+          top4 = top3; top3 = fragcount;
+          idx4 = idx3; idx3 = k;
+        }
+        else if (fragcount > top4)
+        {
+          top4 = fragcount;
+          idx4 = k;
+        }
+      }
+
+      // killough 11/98: replaced cut-and-pasted code with function
+
+      // if the biggest number exists,
+      // put it in the init string
+      i = HU_top(i, idx1, top1);
+
+      // if the second biggest number exists,
+      // put it in the init string
+      i = HU_top(i, idx2, top2);
+
+      // if the third biggest number exists,
+      // put it in the init string
+      i = HU_top(i, idx3, top3);
+
+      // if the fourth biggest number exists,
+      // put it in the init string
+      i = HU_top(i, idx4, top4);
+
+      hud_keysstr[i] = '\0';
+    } //jff 3/17/98 end of deathmatch clause
+    else // build alphabetical key display (not used currently)
```

Comment:
This is currently dead code, but sounds interesting. Maybe we should revive it with a config flag?

Reply:
I think this is an opportunity to delete it 😄

Overall like the changes, I haven’t found any bugs yet.

Reply:
> I think this is an opportunity to delete it 😄

Yep, even better!

> Overall like the changes, I haven’t found any bugs yet.

Excellent! See how handy `HU_Drawer()` has become?

Reply:
> See how handy `HU_Drawer()` has become?

Yes, much better 👍 We can remove this comment now: `// [FG] ~440 lines below`